src/rlookup.c

```diff
@@ -86,8 +53,195 @@ const FieldSpec *findFieldInSpecCache(const RLookup *lookup, const char *name) {
+
+  if (fs) {
+    setKeyByFieldSpec(key, fs);
+  }
+}
+
+static RLookupKey *RLookup_FindKey(RLookup *lookup, const char *name, size_t n) {
+  for (RLookupKey *kk = lookup->head; kk; kk = kk->next) {
+    // match `name` to the name/path of the field
+    if (kk->name_len == n && !strncmp(kk->name, name, kk->name_len)) {
+      return kk;
+    }
+  }
+  return NULL;
+}
+
+RLookupKey *RLookup_GetKeyEx(RLookup *lookup, const char *name, size_t n, RLookupMode mode, int flags) {
+  // remove all flags that are not relevant to getting a key
+  flags &= RLOOKUP_GET_KEY_FLAGS;
+  // First, look for the key in the lookup table for an existing key with the same name
+  RLookupKey *key = RLookup_FindKey(lookup, name, n);
+
+  switch (mode) {
+  // 1. if the key is already loaded, or it has created by earlier RP for writing, return NULL
+  // 2. try generating a key from the schema, including all relevant meta data
+  // 3. if the key is not found in the schema, it might be in the document.
+  //    create a new key with the name of the field, and mark it as doc-source loaded.
+  case RLOOKUP_M_LOAD:
+    if (lookup->spcache == NULL) {
+      // If there is no schema, we can't load the key - fail (error?)
+      return NULL;
+    }
+    if (!key) {
+      key = createNewKey(lookup, name, n);
+    } else if ((key->flags & RLOOKUP_F_ISLOADED || key->flags & RLOOKUP_F_QUERYSRC) && !(flags & RLOOKUP_F_OVERRIDE)) {
+      // If the caller wanted to mark this key as explicit return, mark it as such even if we don't return it.
+      key->flags |= (flags & RLOOKUP_F_EXPLICITRETURN);
+      // Already loaded or created for writing, return NULL (no override)
+      return NULL;
+    }
+    setKeyByFlags(key, flags);
+    setLoadedKey(lookup, key, name, n);
+    return key;
+
+  // A. we found the key at the lookup table:
+  //    1. if we are in exclusive mode, return NULL
+  //    2. if we are in create mode, overwrite the key (remove schema related data, mark with new flags)
+  // B. we didn't find the key at the lookup table:
+  //    create a new key with the name and flags
+  case RLOOKUP_M_WRITE:
+    if (!key) {
+      key = createNewKey(lookup, name, n);
+    } else if (!(flags & RLOOKUP_F_OVERRIDE)) {
+      return NULL;
+    }
+    setKeyByFlags(key, flags);
+    key->flags |= RLOOKUP_F_QUERYSRC;
+    return key;
+
+  // Return the key if it exists in the lookup table, or if it exists in the schema as SORTABLE.
+  case RLOOKUP_M_READ:
+    if (!key) {
+      // If we didn't find the key at the lookup table, check if it exists in
+      // the schema as SORTABLE, and create only if so.
+      key = genKeyFromSpec(lookup, name, n, flags);
+    }
+    // If we didn't find the key in the schema (there is no schema) and unresolved is OK, create an unresolved key.
+    if (!key && (lookup->options & RLOOKUP_OPT_UNRESOLVED_OK)) {
+      key = createNewKey(lookup, name, n);
+      setKeyByFlags(key, flags);
+      key->flags |= RLOOKUP_F_UNRESOLVED;
```

Comment:
is this in use?

Reply:
Yes, by the coordinator, and only for read. In a future PR I'll verify if this is absolutely needed, but I believe it is